torch/csrc/distributed/rpc/rref_context.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include <c10/util/Optional.h>
+#include <torch/csrc/distributed/rpc/message.h>
+#include <torch/csrc/distributed/rpc/rpc_agent.h>
+#include <torch/csrc/distributed/rpc/rref.h>
+#include <torch/csrc/distributed/rpc/types.h>
+
+#include <atomic>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+// Manages RRef lifetime and keeps track of RRef forks.
+class RRefContext {
+ public:
+  static void initInstance(std::shared_ptr<RpcAgent>);
+  static std::unique_ptr<RRefContext>& getInstance();
+
+  RRefContext(const RRefContext &) = delete;
+  void operator=(const RRefContext &) = delete;
+
+  worker_id_t getWorkerId() const;
+  RRefId genRRefId();
+  const std::shared_ptr<RpcAgent>& agent() const;
+
+  // create a new RRef
+  template <typename T>
+  std::shared_ptr<RRef> createRRef(worker_id_t ownerId) {
+    if (ownerId == getWorkerId()) {
+      return getOrCreateOwnerRRef<T>(genRRefId());
+    } else {
+      return createUserRRef<T>(ownerId, genRRefId(), genRRefId());
```

Comment:
For `UserRRef`, we need to call `genRRefId()` twice to get different ids. Because user RRef's `rrefId` and `forkId` should be different. 

But, yes I should've moved `forkId` from `RRef` to `UserRRef`, as `UserRRef` don't need to have a fork id.